magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordsServiceRO.scala

```diff
@@ -205,91 +416,187 @@ class RecordsServiceRO(config: Config, system: ActorSystem, materializer: Materi
+    nickname = "getById",
+    httpMethod = "GET",
+    response = classOf[Record],
+    notes = "Gets a complete record, including data for all aspects."
+  )
+  @ApiImplicitParams(
+    Array(
+      new ApiImplicitParam(
+        name = "id",
+        required = true,
+        dataType = "string",
+        paramType = "path",
+        value = "ID of the record to be fetched."
+      ),
+      new ApiImplicitParam(
+        name = "aspect",
+        required = false,
+        dataType = "string",
+        paramType = "query",
+        allowMultiple = true,
+        value =
+          "The aspects for which to retrieve data, specified as multiple occurrences of this query parameter.  Only records that have all of these aspects will be included in the response."
+      ),
+      new ApiImplicitParam(
+        name = "optionalAspect",
+        required = false,
+        dataType = "string",
+        paramType = "query",
+        allowMultiple = true,
+        value =
+          "The optional aspects for which to retrieve data, specified as multiple occurrences of this query parameter.  These aspects will be included in a record if available, but a record will be included even if it is missing these aspects."
+      ),
+      new ApiImplicitParam(
+        name = "dereference",
+        required = false,
+        dataType = "boolean",
+        paramType = "query",
+        value =
+          "true to automatically dereference links to other records; false to leave them as links.  Dereferencing a link means including the record itself where the link would be.  Dereferencing only happens one level deep, regardless of the value of this parameter."
+      ),
+      new ApiImplicitParam(
+        name = "X-Magda-Tenant-Id",
+        required = true,
+        dataType = "number",
+        paramType = "header",
+        value = "0"
+      )
+    )
+  )
+  @ApiResponses(
+    Array(
+      new ApiResponse(
+        code = 404,
+        message = "No record exists with that ID.",
+        response = classOf[BadRequest]
+      )
+    )
+  )
   def getById: Route = get {
     path(Segment) { id =>
       requiresTenantId { tenantId =>
-        parameters('aspect.*, 'optionalAspect.*, 'dereference.as[Boolean].?) { (aspects, optionalAspects, dereference) =>
-          DB readOnly { session =>
-            recordPersistence.getByIdWithAspects(session, id, tenantId, aspects, optionalAspects, dereference) match {
-              case Some(record) => complete(record)
-              case None => complete(StatusCodes.NotFound, BadRequest("No record exists with that ID or it does not have the required aspects."))
+        parameters('aspect.*, 'optionalAspect.*, 'dereference.as[Boolean].?) {
+          (aspects, optionalAspects, dereference) =>
+            withRecordOpaQuery(AuthOperations.read)(
+              config,
+              system,
+              materializer,
+              ec
+            ) { opaQueries =>
+              assert(opaQueries nonEmpty)
```

Comment:
Strictly speaking it is not necessary. This is a precaution only because authz on write operations are not yet implemented. To make old tests work I skip opa query if opaQueries is empty.